tests : add tests for aggregate cluster circuit breaker

bazmurphy opened the pull request:
Commit Message: tests : add tests for aggregate cluster circuit breaker 
Additional Description: This PR adds tests to the aggregate cluster circuit breaker, to check the circuit breaker behaviour against different scenarios.
Risk Level: Low - tests only
Testing: N/A
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A
Fixes: https://github.com/envoyproxy/envoy/issues/12739

------ 

This PR has two authors: @bazmurphy and @SaadiaELF

Co-authored-by: Saadia El fekak <74792703+SaadiaELF@users.noreply.github.com>
Signed-off-by: Baz Murphy <61154071+bazmurphy@users.noreply.github.com>
Signed-off-by: Saadia El fekak <74792703+SaadiaELF@users.noreply.github.com>

RyanTheOptimist commented:
I think something is confused with this PR as it is show more than 300 changed files. And it definitely needs a main merge.

/wait

bazmurphy commented:
Apologies **RyanTheOptimist** we were trying to fix the DCO problem with a careful edit on our specific commits to make the Author and Signoff the same across each, but it didn't turn out how we expected. I have force pushed back to the original state.

RyanTheOptimist commented:
Great, thanks! Now that `test/per_file_coverage.sh` is not being changed, no coverage shepherd is needed so I'll unassing myself and let the other reviewer take over.

bazmurphy commented:
We were able to fix the DCO sign offs
Seems everything is OK now 🙏

adisuissa reviewed (commented):
Thanks for your contribution!
High-level question: what's the difference between the newly added aggregate-cluster circuit breaker tests and non-aggregated-cluster circuit breaker tests? I may have missed something, but is there anything that's specific for aggregated clusters?
Would it make sense to just have these tests (if they are not already tested) in https://github.com/envoyproxy/envoy/blob/main/test/common/upstream/upstream_impl_test.cc ?

adisuissa commented:
Just to make it clear, I'm for adding tests that validate the unique features of aggregated clusters with circuit breakers. I suggest highlighting those in the tests.

SaadiaELF commented:
1. From this PR we try to prove that the `circuit_breaker` operates only at the level of the aggregate cluster.  As described in this issue https://github.com/envoyproxy/envoy/issues/12739. 
2. lines [**L126-127**](https://github.com/bazmurphy/envoy/blob/897727d4a4a1529f6dc27edf6b3fad6b861d64a6/test/extensions/clusters/aggregate/cluster_test.cc#L126) , show that the `resource_manager` for both child clusters is not created because we don’t have circuit breaker in their level which highlight the unique feature of the aggregate cluster . 
4. The rest of the tests effectively test the behaviour of the circuit breaker using different scenarios of circuit breaking which is common for any type of cluster.
5. The file linked doesn’t test in particular the circuit breaker behaviour. It mainly tests the circuit breaker setup/configuration.

Shall we move the common circuit breaker behaviour to https://github.com/envoyproxy/envoy/blob/main/test/common/upstream/upstream_impl_test.cc ?

adisuissa commented:
> @adisuissa Shall we move the common circuit breaker behaviour to https://github.com/envoyproxy/envoy/blob/main/test/common/upstream/upstream_impl_test.cc ?

SGTM.
And please add a comment describing what the EXPECT_CALLs in the remaining lines are validating.

KBaichoo reviewed (commented):
Thank you for taking on adding tests.

Agree with @adisuissa on just having tests in aggregate cluster that pertain to that features unique interaction with circuit breakers e.g. aggregate circuit breaker on the aggregate cluster trips at value 2 and how the clusters underneath it e.g. with a large number limit doesn't impact the aggregate. Likewise with aggregate having a large value and the underlying clusters having a smaller value shouldn't trip the connection, etc. 

/ wait

bazmurphy commented:
Hello again 😄 

Here's a status update:

We have been continuing to work on this! (so this is not stale)

We have been digging around thoroughly, and found several links (discussions and documents) on the Aggregate Cluster behaviour:

[1] [Discussion](https://github.com/envoyproxy/envoy/issues/13134) [2] [Discussion](https://github.com/envoyproxy/envoy/issues/17412) [3] [Maintainer Doc](https://docs.google.com/document/d/17fc0rAktQpc5oMkerOLHSTnXwX1xWAqW7vZ6_LL47NY/edit?tab=t.0#heading=h.4xxf5tk5hhr0) [4] [gRPC proposal](https://github.com/grpc/proposal/pull/405/files#diff-9a837ca4dc2d15366c277538534e8883f08452ae6a84eac83b87bf98c3767079 ) [5] [the Original PR](https://github.com/envoyproxy/envoy/pull/7967 )

And using what we have learnt from the above, we verified that for ourselves with a live test setup of Envoy observing the circuit breakers and stats.

From all of the above, we then had what we needed to proceed...

So now we are now working on an Integration Test to demonstrate the behaviour we are now more clear about.

And will be back here when we have that ready! 👍 Thanks

bazmurphy commented:
keepalive

bazmurphy commented:
Hello again 😄

We are back and ready for another review

- We have removed the **unit tests**
- We have added **integration tests** for the following aggregate cluster circuit breakers:
  - `max_connections`
  - `max_requests`
  - `max_pending_requests`
  - `max_retries`

- These tests demonstrate the aggregate cluster behaviour discussed [here](https://github.com/envoyproxy/envoy/pull/38251#issuecomment-2704007895)
  - For the `max_connections`, `max_requests`, `max_pending_requests` circuit breakers on the aggregate cluster:
    - When the configured limit is reached on both the aggregate cluster and the underlying cluster(s), only the underlying cluster(s)' circuit breaker opens, the aggregate cluster's circuit breaker is completely unaffected.
    
  - For the `max_retries` circuit breaker on the aggregate cluster:
    - Unlike the other circuit breakers types : 
      - When the configured limit is reached on the aggregate cluster its circuit breaker opens
      - When the configured limit is reached on the underlying cluster its circuit breaker opens
      - Both circuit breakers work independently

Thank you 🙏

markdroth commented:
> For the `max_retries` circuit breaker on the aggregate cluster:
> * Unlike the other circuit breakers types:
>   
>   * When the configured limit is reached on the aggregate cluster its circuit breaker opens
>   * When the configured limit is reached on the underlying cluster its circuit breaker opens
>   * Both circuit breakers work independently

I don't think that's the behavior we discussed in #13134 and #17412.  I think we want the max_retries circuit breaker to work based *only* on the aggregate cluster, not the underlying cluster.

bazmurphy commented:
> I don't think that's the behavior we discussed in #13134 and #17412. I think we want the max_retries circuit breaker to work based _only_ on the aggregate cluster, not the underlying cluster.

Hi @markdroth

Sorry perhaps that part of the comment was not as clear as it could have been, here is another attempt:

In the `max_retries` test that we have added (part of the test shows):
- when a request to the aggregate cluster path needs a retry:
  - the retry only counts against the aggregate cluster's retry budget
  - confirmed by the stats showing the aggregate cluster's circuit breaker opening, whilst the underlying clusters' circuit breaker remains unaffected. specifically, after the first retry:     
    - `cluster.aggregate_cluster.circuit_breakers.default.rq_retry_open` becomes `1` and `remaining_retries` becomes `0`
    - whereas `cluster.cluster_1.circuit_breakers.default.rq_retry_open` remains `0` and `remaining_retries` stays at `1`, despite cluster_1 actually handling the retry

So when the aggregate cluster's retry circuit breaker opens, subsequent requests to the aggregate cluster path cannot retry, even though the underlying cluster's retry budget is still available.

markdroth commented:
Okay, that sounds right, then.  Sorry for misunderstanding!

KBaichoo commented on test/extensions/clusters/aggregate/cluster_integration_test.cc:
```diff
@@ -178,6 +211,50 @@ class AggregateIntegrationTest
+
+    auto* cluster_circuit_breakers_threshold_default = cluster_circuit_breakers->add_thresholds();
+    cluster_circuit_breakers_threshold_default->set_priority(
+        envoy::config::core::v3::RoutingPriority::DEFAULT);
+
+    cluster_circuit_breakers_threshold_default->mutable_max_connections()->set_value(
+        limits.max_connections);
+    cluster_circuit_breakers_threshold_default->mutable_max_pending_requests()->set_value(
+        limits.max_pending_requests);
+    cluster_circuit_breakers_threshold_default->mutable_max_requests()->set_value(
+        limits.max_requests);
+    cluster_circuit_breakers_threshold_default->mutable_max_retries()->set_value(
+        limits.max_retries);
+    cluster_circuit_breakers_threshold_default->mutable_max_connection_pools()->set_value(
+        limits.max_connection_pools);
+    cluster_circuit_breakers_threshold_default->set_track_remaining(true);
+  }
+
+  void setMaxConcurrentStreams(envoy::config::cluster::v3::Cluster& cluster,
+                               uint32_t max_concurrent_streams) {
+    envoy::extensions::upstreams::http::v3::HttpProtocolOptions http_protocol_options;
+    http_protocol_options.mutable_explicit_http_config()
+        ->mutable_http2_protocol_options()
+        ->mutable_max_concurrent_streams()
+        ->set_value(max_concurrent_streams);
+    (*cluster.mutable_typed_extension_protocol_options())
+        ["envoy.extensions.upstreams.http.v3.HttpProtocolOptions"]
+            .PackFrom(http_protocol_options);
+  }
+
+  void
+  reduceAggregateClustersListToOneCluster(envoy::config::cluster::v3::Cluster& aggregate_cluster) {
+    auto* aggregate_cluster_type = aggregate_cluster.mutable_cluster_type();
+    auto* aggregate_cluster_typed_config = aggregate_cluster_type->mutable_typed_config();
+    envoy::extensions::clusters::aggregate::v3::ClusterConfig new_aggregate_cluster_typed_config;
+    aggregate_cluster_typed_config->UnpackTo(&new_aggregate_cluster_typed_config);
+    new_aggregate_cluster_typed_config.clear_clusters();
+    new_aggregate_cluster_typed_config.add_clusters("cluster_1");
+    aggregate_cluster_typed_config->PackFrom(new_aggregate_cluster_typed_config);
+  }
```
These call all be free functions in an anonymous namespace since they don't rely on any of the members of the test fixture.

KBaichoo reviewed (commented):
Thank you for the extensive addition of rigorous test for this behavior. Can you add a section to:
https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/aggregate_cluster to specify
the behavior of circuit breakers for aggregate clusters. It can be a bit unintuitive and end-users should not have to read these test to figure out how this works.

LGTM otherwise besides the minor comments.

/wait

SaadiaELF commented:
Hello again 👋 ,

As requested we have : 
Addressed the minor comments.
Added a section on aggregate cluster circuit breaker behaviour to the documentation.

Thank you 🙇‍♀️ !

KBaichoo reviewed (approved):
:tada: Thank you!
